cobbler/actions/buildiso/__init__.py

```diff
@@ -0,0 +1,251 @@
+"""
+Builds bootable CD images that have PXE-equivalent behavior for all Cobbler distros/profiles/systems currently in
+memory.
+"""
+
+# SPDX-License-Identifier: GPL-2.0-or-later
+# SPDX-FileCopyrightText: Copyright 2006-2009, Red Hat, Inc and Others
+# SPDX-FileCopyrightText: Michael DeHaan <michael.dehaan AT gmail>
+
+import logging
+import os
+import os.path
+import pathlib
+import shutil
+from typing import Dict, List, Optional, Union
+
+from cobbler import utils
+
+
+def add_remaining_kopts(koptdict: dict) -> str:
+    """Add remaining kernel_options to append_line
+
+    :param koptdict: The kernel options which are not present in append_line.
+    :return: A single line with all kernel options from the dictionary in the string. Starts with a space.
+    """
+    append_line = ""
+    for key, value in koptdict.items():
+        if value is None:
+            append_line += " %s" % key
+        else:
+            if isinstance(value, list):
+                for i in value:
+                    _i = str(i).strip()
+                    if ' ' in _i:
+                        append_line += " %s='%s'" % (key, _i)
+                    else:
+                        append_line += " %s=%s" % (key, _i)
+            else:
+                _v = str(value).strip()
+                if ' ' in _v:
+                    append_line += " %s='%s'" % (key, _v)
+                else:
+                    append_line += " %s=%s" % (key, _v)
+    return append_line
```

Comment:
IIRC you only moved this around, I don't know if you want to improve it at this time or later on. I see two main points I would change here.

1. Building strings with `+=` which copies the strings (actually, CPython does optimize for this very common anti-pattern and can re-use strings, but I'd use the correct pattern nonetheless)
2. Repetition in branches, can be re-ordered a bit

This is is the first idea I came up , I only tested it with dummy content I made up:
```python
def add_remaining_kopts(kopts: dict) -> str:
    append_line = [""] # empty str to ensure the returned str starts with a space
    for option, args in kopts.items():
        if args is None:
            append_line.append(f"{option}")
            continue

        if not isinstance(args, list):
            args = [args]

        for arg in args:
            arg_str = format(arg)
            if ' ' in arg_str:
                arg_str = f'"{arg_str}"'
            append_line.append(f"{option}={arg_str}")

    return ' '.join(append_line)
```

Reply:
I took your suggestion over. We have a test for this and we should extend it, so we can be sure that this works in the different scenarios.